### PR TITLE
fix qasm to yao translation

### PR DESCRIPTION
**Pull request by GiggleLiu**
## Changes
* new yao APIs `MEASURE`, `RESET` and `Barrier`
* fix qreg address translation
* fix function translation

## Notes
* the 3rd parameter of `MEASURE` should be a view of classical register, like `view(c, [2,3])`, is this possible?
* one of your test is broken. @thautwarm

### CODE REVIEW

**Comment by GiggleLiu** [comment]
Now the following example works perfectly, isn't this amazing @Roger-luo  @wangleiphy 

```julia
julia> src1 = """
       OPENQASM 2.0;

       gate cu1(theta) a,b
       {
           U(0,0,theta/2) a;
           CX a,b;
           U(0,0,-theta/2) b;
       }

       qreg q[3];
       qreg a[2];
       creg c[3];
       creg syn[2];
       cu1(pi/2) q[0],q[1];
       """
"OPENQASM 2.0;\n\ngate cu1(theta) a,b\n{\n    U(0,0,theta/2) a;\n    CX a,b;\n    U(0,0,-theta/2) b;\n}\n\nqreg q[3];\nqreg a[2];\ncreg c[3];\ncreg syn[2];\ncu1(pi/2) q[0],q[1];\n"

julia> circ = @eval $(trans(parse_qasm(lex(src1))))
nqubits: 5
chain
├─ put on (1)
│  └─ chain
│     ├─ rot(Z gate, 0.7853981633974483)
│     ├─ rot(Y gate, 0)
│     └─ rot(Z gate, 0)
├─ control(1)
│  └─ (2,) X gate
└─ put on (2)
   └─ chain
      ├─ rot(Z gate, -0.7853981633974483)
      ├─ rot(Y gate, 0)
      └─ rot(Z gate, 0)
```

**Comment by Roger-luo** [comment]
can you fix the CI? @GiggleLiu the PR looks good in general. I'd like to merge it and register the package first.

File: src/Bag.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# This file is copied from YaoExtensions for convenience,
+# will be considerd to move to YaoBlocks!
+export AbstractBag
+"""
+    AbstractBag{BT, N}<:TagBlock{BT, N}
```
Comment (Roger-luo) [inline src/Bag.jl:5]: Isn't this in YaoExtension or some early version of YaoBlocks already? it looks a bit messy now.

Comment (GiggleLiu) [reply src/Bag.jl:5]: yes，it turns out this wrapper class is pretty useful, it keeps all property of content (more strict than TagBlock). should be considered to moved back to YaoBlocks.

File: src/Bag.jl [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+abstract type AbstractBag{BT, N}<:TagBlock{BT, N} end
+
+YaoBlocks.mat(::Type{T}, bag::AbstractBag{N}) where {T,N} = mat(T, content(bag))
+YaoBlocks.apply!(reg::AbstractRegister, bag::AbstractBag) = apply!(reg, content(bag))
+YaoBlocks.ishermitian(bag::AbstractBag) = ishermitian(content(bag))
+YaoBlocks.isreflexive(bag::AbstractBag) = isreflexive(content(bag))
+YaoBlocks.isunitary(bag::AbstractBag) = isunitary(content(bag))
+YaoBlocks.occupied_locs(bag::AbstractBag) = occupied_locs(content(bag))
+
+function Yao.AD.apply_back!(state, b::AbstractBag, collector)
```
Comment (Roger-luo) [inline src/Bag.jl:20]: `YaoBlocks.AD` this is a small component package,  it shouldn't depend on Yao (the meta package).

Comment (Roger-luo) [reply src/Bag.jl:20]: Can you fix this so the CI passes? @GiggleLiu

Comment (GiggleLiu) [reply src/Bag.jl:20]: I will break the test for the momentum since I don't know how to fix. @thautwarm Do you have time to look into it?

**Pull request merged**